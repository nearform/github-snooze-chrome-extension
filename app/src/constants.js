/**
 * Prefixes
 */
export const GH_SNOOZE_PREFIX = 'GH_SNOOZE - '

/**
 * Errors
 */
export const ERR_USER_NOT_FOUND = 'User not found.'

/**
 * Messages
 */
export const MSG_URL_CHANGED = 'URL_CHANGED'
export const MSG_TAB_REFRESHED = 'TAB_REFRESHED'
