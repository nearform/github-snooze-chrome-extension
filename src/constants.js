/**
 * Messages
 */
export const MSG_TAB_REFRESHED = 'MSG_TAB_REFRESHED'
export const MSG_URL_CHANGED = 'MSG_URL_CHANGED'

/**
 * Snooze Statuses
 */
export const SNOOZE_STATUS_PENDING = 'SNOOZE_STATUS_PENDING'
export const SNOOZE_STATUS_DONE = 'SNOOZE_STATUS_DONE'

/**
 * Entity Types
 */
export const ENTITY_TYPE_ISSUE = 'ENTITY_TYPE_ISSUE'
export const ENTITY_TYPE_PR = 'ENTITY_TYPE_PR'

/**
 * Actions
 */
export const ACTION_UPDATE_BADGE_COUNTER = 'ACTION_UPDATE_BADGE_COUNTER'

/**
 * URL
 */
export const URL_MATCH = 'https://github.com/'

/**
 * Colors
 */
export const COLOR_PRIMARY = '#140048'
export const COLOR_SECONDARY = '#FB7A9C'
export const COLOR_SUCCESS = '#E3FBE3'
export const COLOR_WHITE = '#FFFFFF'

/**
 * Storage Keys
 */
export const SK_BADGE_COUNTER = 'badgeCounter'
export const SK_URL = 'url'
export const SK_USER = 'user'
export const SK_PAT = 'pat'

/**
 * Reducer Action Types
 */
export const SET_LOADING = 'SET_LOADING'
export const SET_AUTHENTICATED = 'SET_AUTHENTICATED'
export const SET_USER = 'SET_USER'
export const SET_PAT = 'SET_PAT'
export const SET_CURRENT_URL = 'SET_CURRENT_URL'
export const SET_SNOOZE_LIST = 'SET_SNOOZE_LIST'
