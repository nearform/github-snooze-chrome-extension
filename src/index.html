<!DOCTYPE html>
<html lang="en">
  <head>
    <link
      rel="stylesheet"
      href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"
    />
    <link
      rel="stylesheet"
      href="https://fonts.googleapis.com/icon?family=Material+Icons"
    />
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=1000, initial-scale=1.0" />
    <title>Popup Page</title>
  </head>
  <body style="margin: 0; padding: 0; min-height: 500px; max-height: 600px;">
    <div id="root"></div>
    <script src="index.js"></script>
  </body>
</html>
