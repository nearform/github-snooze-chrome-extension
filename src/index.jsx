import React from 'react'
import { createRoot } from 'react-dom/client'
import { readFromLocalStorage } from './api/chrome'
import { createClient, Provider } from 'urql';
import {
    SK_PAT,
    SK_USER,
  } from './constants'

import App from './App'

const client = createClient({
    url: 'https://api.github.com/graphql',
    fetchOptions: async () => {
        const { pat } =  await readFromLocalStorage([SK_PAT, SK_USER])
        
        if (!pat) {
            throw Error('PAT not available.')
        }
        return {
            headers: { authorization: `Bearer ${pat}` },
        };
    },
});

const root = createRoot(document.getElementById('root'))

root.render(
    <Provider value={client}>
        <App />
    </Provider>,
)
