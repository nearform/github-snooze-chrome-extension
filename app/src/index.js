import { ERR_USER_NOT_FOUND } from './constants.js'
import { logDebug, logError, logInfo } from './logger.js'
import { fetchUserData, fetchUsername, initUserModel } from './user.js'

const start = async () => {
  // Retrieve the logged in user from the meta user-login
  const username = fetchUsername()
  if (!username) {
    logError(ERR_USER_NOT_FOUND)
    return
  }
  logInfo('User found: ' + username)

  // Retrieve user data from GitHub API
  const userData = await fetchUserData(username)
  if (!userData) {
    logError(`User data for @${username} not found.`)
  }

  // Initialise the user model with needed data
  const user = initUserModel(userData)
  logDebug('User initialised: ' + user)
}

start()
