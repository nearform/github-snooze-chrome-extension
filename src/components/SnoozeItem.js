import React from 'react'
import {
  CardHeader,
  Link,
  Box,
  CardContent,
  Typography,
  useTheme,
  IconButton
} from '@mui/material'
import {
  Business as BusinessIcon,
  Cancel as CancelIcon
} from '@mui/icons-material'
import DialogButton from './DialogButton'
import { dateHasPassed, getFormattedDate } from '../date'
import { SNOOZE_STATUS_DONE } from '../constants'
import { removeSnooze } from '../api/chrome'
import SnoozeCard from './SnoozeCard'
import { getEntityInfo } from '../parser'
import { SnoozeShape } from '../shapes'

function SnoozeItem({ user, snooze, onDelete }) {
  const theme = useTheme()

  const { url, notifyAt, status } = snooze
  const { owner } = getEntityInfo(url)

  const handleDelete = async () => {
    const updatedSnoozeList = await removeSnooze(user.id, snooze)
    onDelete(updatedSnoozeList)
  }

  const hasPassed = dateHasPassed(notifyAt, new Date().getTime())
  const notifyLabel = hasPassed ? 'Notified at': 'Scheduled for'

  return (
    <SnoozeCard square elevation={0} sx={{ py: 0.5 }} status={status}>
      <CardHeader
        sx={{ py: 0 }}
        subheader={
          <Box sx={{ display: 'flex', flexDirection: 'row' }}>
            <BusinessIcon />
            <Box sx={{ ml: 0.5 }}>
              <Typography
                // sx={{ lineHeight: 1.5 }}
                fontWeight="bold"
                variant="subtitle1"
                color={theme.palette.secondaryLightest.main}
              >
                {owner}
              </Typography>
            </Box>
          </Box>
        }
      />
      <CardContent
        sx={{
          ':last-child': {
            pb: 0.5,
            pt: 0
          }
        }}
      >
        <Box
          sx={{
            display: 'flex',
            justifyContent: 'space-between',
            alignItems: 'center'
          }}
        >
          <Link
            color={theme.palette.secondary.main}
            href={url}
            variant="body1"
            target="_blank"
          >
            {url}
          </Link>
          {status === SNOOZE_STATUS_DONE ? (
            <IconButton color="secondary" onClick={handleDelete}>
              <CancelIcon />
            </IconButton>
          ) : (
            <DialogButton
              icon={<CancelIcon />}
              title="Attention"
              description="Do you want to delete this snooze?"
              onConfirm={handleDelete}
            />
          )}
        </Box>
        <Typography
          color={theme.palette.secondaryLight.main}
          variant="subtitle1"
        >
          {notifyLabel} {getFormattedDate(notifyAt)}
        </Typography>
      </CardContent>
    </SnoozeCard>
  )
}

SnoozeItem.propTypes = {
  snooze: SnoozeShape.isRequired
}

export default SnoozeItem
