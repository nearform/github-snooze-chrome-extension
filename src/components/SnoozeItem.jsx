import React from 'react'
import {
  CardHeader,
  Link,
  Box,
  CardContent,
  Typography,
  useTheme,
  IconButton
} from '@mui/material'
import {
  Business as BusinessIcon,
  Cancel as CancelIcon
} from '@mui/icons-material'
import DialogButton from './DialogButton'
import { SNOOZE_STATUS_DONE } from '../constants'
import { removeSnooze } from '../api/chrome'
import SnoozeCard from './SnoozeCard'
import { getEntityInfo } from '../parser'
import { SnoozeShape } from '../shapes'
import SnoozeItemText from './SnoozeItemText'

function SnoozeItem({ user, snooze, onDelete }) {
  const theme = useTheme()

  const { url, notifyAt, status } = snooze
  const { owner } = getEntityInfo(url)

  const handleDelete = async () => {
    const updatedSnoozeList = await removeSnooze(user.id, snooze)
    onDelete(updatedSnoozeList)
  }

  return (
    <SnoozeCard square elevation={0} sx={{ py: 0.5 }} status={status}>
      <CardHeader
        sx={{ py: 0 }}
        subheader={
          <Box sx={{ display: 'flex', flexDirection: 'row' }}>
            <BusinessIcon />
            <Box sx={{ ml: 0.5 }}>
              <Typography
                // sx={{ lineHeight: 1.5 }}
                fontWeight="bold"
                variant="subtitle1"
                color={theme.palette.secondaryLightest.main}
              >
                {owner}
              </Typography>
            </Box>
          </Box>
        }
      />
      <CardContent
        sx={{
          ':last-child': {
            pb: 0.5,
            pt: 0
          }
        }}
      >
        <Box
          sx={{
            display: 'flex',
            justifyContent: 'space-between',
            alignItems: 'center'
          }}
        >
          <Link
            color={theme.palette.secondary.main}
            href={url}
            variant="body1"
            target="_blank"
          >
            {url}
          </Link>
          {status === SNOOZE_STATUS_DONE ? (
            <IconButton color="secondary" onClick={handleDelete}>
              <CancelIcon />
            </IconButton>
          ) : (
            <DialogButton
              icon={<CancelIcon />}
              title="Attention"
              description="Do you want to delete this snooze?"
              onConfirm={handleDelete}
            />
          )}
        </Box>
        <Typography
          color={theme.palette.secondaryLight.main}
          variant="subtitle1"
        >
          <SnoozeItemText notifyAt={notifyAt} />
        </Typography>
      </CardContent>
    </SnoozeCard>
  )
}

SnoozeItem.propTypes = {
  snooze: SnoozeShape.isRequired
}

export default SnoozeItem
