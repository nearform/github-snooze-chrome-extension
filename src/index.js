import React from 'react'
import { createRoot } from 'react-dom/client'
import { MemoryRouter as Router } from 'react-router-dom'
import { ThemeProvider, createTheme } from '@mui/material/styles'
import App from './App'
import { COLOR_PRIMARY, COLOR_SECONDARY } from './constants'

const theme = createTheme({
  palette: {
    primary: {
      main: COLOR_PRIMARY
    },
    secondary: {
      main: COLOR_SECONDARY
    }
  }
})

const root = createRoot(document.getElementById('root'))

root.render(
  <ThemeProvider theme={theme}>
    <Router>
      <App />
    </Router>
  </ThemeProvider>
)
