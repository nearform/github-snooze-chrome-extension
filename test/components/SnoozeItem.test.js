import React from 'react'
import { render } from '../renderer'
import SnoozeItem from '../../src/components/SnoozeItem'
import { SNOOZE_STATUS_PENDING } from '../../src/constants'

const ISO_STRING_DATE_BASE = '2022-07-28T10:00:00.000Z'

const props = {
  user: {
    id: 1
  },
  snooze: {
    url: 'https://github.com/owner/repo/pull/7',
    notifyAt: new Date(ISO_STRING_DATE_BASE).getTime(),
    status: SNOOZE_STATUS_PENDING
  },
  onDelete: jest.fn()
}

describe('SnoozeItem', () => {
  test('shows the proper SnoozeItem', async () => {
    const { asFragment } = render(<SnoozeItem {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })
})
