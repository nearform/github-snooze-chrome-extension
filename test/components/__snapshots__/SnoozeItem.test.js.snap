// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnoozeItem shows the proper SnoozeItem 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiCard-root css-tu5f03-MuiPaper-root-MuiCard-root"
    status="SNOOZE_STATUS_PENDING"
  >
    <div
      class="MuiCardHeader-root css-rqthld-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiCardHeader-subheader css-1xt7grh-MuiTypography-root"
        >
          <div
            class="MuiBox-root css-1xhj18k"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="BusinessIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 7V3H2v18h20V7H12zM6 19H4v-2h2v2zm0-4H4v-2h2v2zm0-4H4V9h2v2zm0-4H4V5h2v2zm4 12H8v-2h2v2zm0-4H8v-2h2v2zm0-4H8V9h2v2zm0-4H8V5h2v2zm10 12h-8v-2h2v-2h-2v-2h2v-2h-2V9h8v10zm-2-8h-2v2h2v-2zm0 4h-2v2h2v-2z"
              />
            </svg>
            <div
              class="MuiBox-root css-15fzge"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1ayqy7m-MuiTypography-root"
              >
                owner
              </h6>
            </div>
          </div>
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-1yehjro-MuiCardContent-root"
    >
      <div
        class="MuiBox-root css-69i1ev"
      >
        <a
          class="MuiTypography-root MuiTypography-body1 MuiLink-root MuiLink-underlineAlways css-u4a50l-MuiTypography-root-MuiLink-root"
          href="https://github.com/owner/repo/pull/7"
          target="_blank"
        >
          https://github.com/owner/repo/pull/7
        </a>
        <div>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-ynscgk-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CancelIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <h6
        class="MuiTypography-root MuiTypography-subtitle1 css-i5mk92-MuiTypography-root"
      >
        Scheduled for Jul 28 2022 10:00
      </h6>
    </div>
  </div>
</DocumentFragment>
`;
